### PR TITLE
Double-factorized QPE

### PR DESCRIPTION
**Pull request by max-radin**
## Description

While benchq supports coarse resource estimates for single-factorized QPE, double factorization is known to require much less quantum resources. This PR adds functionality for getting resource estimates for double-factorized QPE based on OpenFermion's implementation.

The example below shows how single and double factorized resource estimates can be obtained. Note that this PR does contain a breaking change: the `get_single_factorized_qpe_resource_estimate` has been removed in order to make a clear separation between getting gate/logical-qubit counts and the physical costing.

```python
from benchq.resource_estimation.openfermion_re import (
    get_single_factorized_qpe_toffoli_and_qubit_cost,
    get_double_factorized_qpe_toffoli_and_qubit_cost,
    get_physical_cost
)
from benchq.problem_ingestion.molecule_instance_generation import generate_hydrogen_chain_instance
from openfermion.resource_estimates.molecule import pyscf_to_cas

# Get the Hamiltonian
instance = generate_hydrogen_chain_instance(8)
h1, eri_full, _, _, _ = pyscf_to_cas(instance.get_active_space_meanfield_object())

# Get the number of Toffoli gates and qubits for single and double factorized QPE
num_toffoli_sf, num_qubits_sf = get_single_factorized_qpe_toffoli_and_qubit_cost(h1, eri_full, 20)
num_toffoli_df, num_qubits_df = get_double_factorized_qpe_toffoli_and_qubit_cost(h1, eri_full, 1e-6)

# Get the physical cost
resource_estimates_sf = get_physical_cost(num_toffoli_sf, num_qubits_sf)
resource_estimates_df = get_physical_cost(num_toffoli_df, num_qubits_df)
```

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [x] I have updated documentation.

### CODE REVIEW

**Review by AthenaCaesura — APPROVED** [review]
Approving this for now just so that we can merge later it. But I would highly suggest waiting on the approval of people who actually know what this code does. 😅

**Review by mkodrycka — APPROVED** [review]
Thank you for this PR! It looks really great.

File: src/benchq/resource_estimation/_compute_lambda.py
Comment (max-radin) [reply src/benchq/resource_estimation/_compute_lambda.py]: Good catch! Fixed in 8b183f99d364676b23e8781e1a29f7e7eafad523

**Pull request merged**
